src/buffer.rs

```diff
@@ -362,12 +389,72 @@ impl Buffer {
         }
         self.area = area;
     }
+
+    /// Builds a minimal sequence of coordinates and Cells necessary to update the UI from
+    /// self to other.
+    ///
+    /// We're assuming that buffers are well-formed, that is no double-width cell is followed by
+    /// a non-blank cell.
+    ///
+    /// # Multi-width characters handling:
+    ///
+    /// ```text
+    /// (Index:) `01`
+    /// Prev:    `コ`
+    /// Next:    `aa`
+    /// Updates: `0: a, 1: a'
+    /// ```
+    ///
+    /// ```text
+    /// (Index:) `01`
+    /// Prev:    `a `
+    /// Next:    `コ`
+    /// Updates: `0: コ` (double width symbol at index 0 - skip index 1)
+    /// ```
+    ///
+    /// ```text
+    /// (Index:) `012`
+    /// Prev:    `aaa`
+    /// Next:    `aコ`
+    /// Updates: `0: a, 1: コ` (double width symbol at index 1 - skip index 2)
+    /// ```
+    pub fn diff<'a>(&self, other: &'a Buffer) -> Vec<(u16, u16, &'a Cell)> {
```

Comment:
The previous algorithm would be enough if we could guarantee that all phantom cells (extra cells occupied by a multi-width char) are empty (`symbol: ""`), while regular empty cells are blank (`symbol: " "`), which would make them different. This would require care in every place text is rendered - always using either `set_string`, `set_stringn`, or some other method which takes care of variable symbol width, never inserting a symbol manually with `buf.get_mut().set_symbol()`.

Right now Paragraph happens to produce a well-formed buffer despite (`buf.get_mut().set_symbol()`) because the background is always redrawn.